client/galaxy/scripts/mvc/workflow/workflow-terminals.js

```diff
@@ -370,6 +377,12 @@ var BaseInputTerminal = Terminal.extend({
             `Effective output data type(s) [${cat_outputs}] do not appear to match input type(s) [${this.datatypes}].`
         );
     },
+    _producesAcceptableDatatypeAndOptionalness(other) {
+        if (!this.optional && !this.multiple && other.optional) {
```

Comment:
I guess that will work if a tool author properly handles the iteration, but in my tests I used the cat tool in https://toolshed.g2.bx.psu.edu/repository?repository_id=2593fd36ae8011aa which does
```
#echo ' '.join(['"%s"' % $file for $file in $inputs])#
```
which evaluates to `"None"`. I guess the proper way is `#echo ' '.join(['"%s"' % $file for $file in $inputs if $file])#` or make sure that an empty multi-data parameter behaves like `[]`.
The tool form doesn't let you do this and insists on passing a dataset ... thought I can see this being useful.
So I guess this is fine for now and we should see if it is possible to fix this style of iteration on the Galaxy side ?

Reply:
The tool form insist on a dataset even if it is marked as optional? I thought I hacked around this years ago for Dan... ugh

Reply:
it's the other way round, this code allows connecting optional datasets to required multi-data parameters. I think there is some merit in not allowing that, as the tool form does.

Reply:
Oh - that is 100% intentional on my part. You should be able to connect two optional data parameters to a required multi-parameter - in case one of them is set, right? The tool for lets you do this - run two tools and pick the output that is produced manually and send it 😃. This should be caught at runtime IMO so you can do conditionally things